From 0d8c4acac0aa874750abf07c9555905ab88e1773 Mon Sep 17 00:00:00 2001
From: Marc Mutz <marc.mutz@qt.io>
Date: Sun, 7 Sep 2025 19:36:14 +0200
Subject: [PATCH 045/553] QLocale: enable NRVO in matchingLocales()

The two guard clauses returned temporaries, so enjoyed RVO, but in
doing so destroyed NRVO for the remainder of the function, at least
for most compilers: While the standard allows NRVO even for the old
code (it's statically clear which of the three possible objects gets
returned, before any of them are being created), experience shows that
compilers don't enable NRVO in these situations.

So reformulate the function to be NRVO-friendly: Define the return
object as the first object in the function, and have all return
statements only return that same object. This, on all known compilers,
enables NRVO.

As a drive-by, use emplace_back() instead of append(rvalue).

Amends the start of the public history.

Pick-to: 6.9 6.8
Change-Id: Ie526033f026b313f072912c8f43fa7fe2a1b2b73
Reviewed-by: Edward Welbourne <edward.welbourne@qt.io>
(cherry picked from commit 22a21fc4c6bf8c12e813de77e28f56aaf64b5686)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/corelib/text/qlocale.cpp | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/src/corelib/text/qlocale.cpp b/src/corelib/text/qlocale.cpp
index 1a15cb4311f..1ea2829a3db 100644
--- a/src/corelib/text/qlocale.cpp
+++ b/src/corelib/text/qlocale.cpp
@@ -3155,14 +3155,17 @@ QLocale QLocale::system()
 */
 QList<QLocale> QLocale::matchingLocales(Language language, Script script, Territory territory)
 {
+    QList<QLocale> result;
+
     const QLocaleId filter { language, script, territory };
     if (!filter.isValid())
-        return QList<QLocale>();
+        return result;
 
-    if (language == C)
-        return QList<QLocale>{QLocale(C)};
+    if (language == C) {
+        result.emplace_back(C);
+        return result;
+    }
 
-    QList<QLocale> result;
     if (filter.matchesAll())
         result.reserve(locale_data_size);
 
-- 
2.51.2

